### ISSUE TITLE
boxes format in calculation of huber_loss

### ISSUE THREAD

**Issue by zlyin** [rwightman/efficientdet-pytorch#58]
Hi @rwightman, I'm trying to implement a custom iou loss function. But I'd like to confirm with you about the boxes format consumed by `huber_loss` function. Could you help me verify the format of inputs & targets args?

```
def huber_loss(input, target, delta: float = 1., weights: Optional[torch.Tensor] = None, size_average: bool = True):
    """
    """
    err = input - target
    abs_err = err.abs()
    quadratic = torch.clamp(abs_err, max=delta)
    linear = abs_err - quadratic
    loss = 0.5 * quadratic.pow(2) + delta * linear
    if weights is not None:
        loss *= weights
    return loss.mean() if size_average else loss.sum()
```
I print both of them out and found they are in the shape of `[batch, height_l, width_l, 9*4]`, the last dim of which I think coords for bounding boxes. In other threads, you mentioned that you implementation consumes `targets in YXYX format`, outputs `pred boxes in XYWH format`. Does such theory hold here as well?

Thank you for your confirmation!

**Comment by rwightman** [comment]
Yes, they should be relative (to anchors) regression coordinates in yxyx, they are not converted into absolute xyxy until `decode_box_outputs(boxes, anchors, output_xyxy=True)` is called 
https://github.com/rwightman/efficientdet-pytorch/blob/20cd5f3b9744a5939534ac1c556689ff4a2edd12/effdet/anchors.py#L48-L82) 

... which is usually only done for predictions via the `generate_detections()` call

Note there is a PR for IOU loss at https://github.com/rwightman/efficientdet-pytorch/pull/52 ... I haven't dug in yet but am hoping to take a closer look this week, maybe possible to optimize a bit

**Comment by zlyin** [comment]
Hi @rwightman Great thanks for your quick reply. I implemented a version yesterday but messed up the boxes format, which led to NaN value of loss. Now I can try to fix it. Thanks!

**Comment by zlyin** [comment]
@rwightman While I still have you here, let me ask another question. I'm also trying to modify the default anchors but have no idea of the right way to set it up. Maybe I misunderstood the parameters relevant to anchors in the `model_config.py` file. 

```
    # feature + anchor config
    h.min_level = 3
    h.max_level = 7
    h.num_levels = h.max_level - h.min_level + 1
    h.num_scales = 3
    h.aspect_ratios = [(1.0, 1.0), (1.4, 0.7), (0.7, 1.4)]
    h.anchor_scale = 4.0
```

As you indicated in your code, your implementation employed `3 aspect ratios x 3 octave_scale`, 9 anchors per location on each feature map. I'm following the thread and try to set, for example, `h.aspect_ratios = [(1.0, 1.0), (1.0, 1.5), (1.5, 1.0), (1.0, 2.0), (2.0, 1.0)] `.

**Comment by rwightman** [comment]
To change the anchor sizes anchor_scale is the main factor, it can be defined as a list so it can change per feature level, it's multiplied by the feat stride at different levels, but also multiplied by the aspects (so a 2,1 vs 1.4, .07 * the scale would result in different size anchors)... as for more detail than that you'll have to dig in, the code is adapted from the Google TF models retinanet and other anchor based models, I have not explored the space of options.

**Comment by zlyin** [comment]
Hi Ross, thanks for your explanation. I'll dig into it further. Thank you for your help!
